### PR TITLE
support for IWebJobsConfigurationStartup

### PR DESCRIPTION
**Pull request by brettsam**
Needs tests, but creating draft for discussion.

### CODE REVIEW

**Comment by jamesharling** [comment]
Would it be possible to get access to the constructed `IConfiguration` instance in `Configure(IFunctionsHostBuilder builder)`? This saves users having to build the service provider themselves just to get hold of configuration. This is commonly seen in things like ASP.NET Core, where methods like `ConfigureServices` have overloads exposing `HostBuilderContext` (which exposes config), or you can constructor inject `IConfiguration` in to your startup class.

I would propose changing the signature of `Configure(IFunctionsHostBuilder builder)` to `Configure(IFunctionsHostBuilder builder, FunctionsHostBuilderContext context)` (breaking change) or extending `IFunctionsHostBuilder` with a `Context` or `Configuration` property.

**Comment by brettsam** [comment]
@jamesharling -- yes, I think we need to do that as well. We can follow up with that as a separate PR as there's some additional considerations to think through.

**Comment by pherbel** [comment]
@brettsam 
What do you think?
When can it be released this official support for Configuration flexibility?

**Comment by idg10** [comment]
I'd also like to see something like @jamesharling 's suggestion. The problem with `IConfiguration` being available only via DI is that it's quite common to want to have access to the configuration _during_ DI initialization.

For example, take the Options mechanism—it offers loads of extension methods for describing how you want your configuration to map onto `IOptions<T>` instances, but the majority expect you to have an `IConfiguration` in hand.

By making `IConfiguration` available only through DI, you end up with a lot of awkward callback-based DI registrations where you have to use the `IServiceProvider` in the callback to obtain the configuration.

**Comment by brettsam** [comment]
Latest PR exposes a context that allows you access to the IConfiguration (see the sample changes).

File: src/Extensions/DependencyInjection/FunctionsHostBuilderContext.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+        public FunctionsHostBuilderContext(WebJobsBuilderContext webJobsBuilderContext)
+        {
+            _webJobsBuilderContext = webJobsBuilderContext ?? throw new ArgumentNullException(nameof(webJobsBuilderContext));
+        }
+
+        /// <summary>
+        /// Gets the name of the environment. This is "Development" when running in the Azure Functions Core Tools and "Production" when deployed to Azure.
+        /// </summary>
+        public string EnvironmentName => _webJobsBuilderContext.EnvironmentName;
```
Comment (jamesharling) [inline src/Extensions/DependencyInjection/FunctionsHostBuilderContext.cs:25]: Would it be possible to expose `IHostEnvironment` here (which in turn exposes the EnvironmentName)? This gives parity with the generic host, whilst also being a little more flexible.

Comment (brettsam) [reply src/Extensions/DependencyInjection/FunctionsHostBuilderContext.cs:25]: That introduced a little more complexity. The `IHostEnvironment` interface changed between .NET Core 2 and 3 so we'd need to multi-target this project and deal with those differences. We decided to abstract that away and not directly expose the interface. The other properties were also confusing in the context of Functions as they refer to the location and name of the host on the VM, not your deployed application.

This value is exactly the `IHostEnvironment.EnvironmentName` -- it gets copied here by the host. If there are other values from `IHostEnvironment` that are useful, we can expose them here.

File: src/Extensions/DependencyInjection/FunctionsConfigurationBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+{
+    internal class FunctionsConfigurationBuilder : IFunctionsConfigurationBuilder, IFunctionsHostBuilderExt
+    {
+        public FunctionsConfigurationBuilder(IConfigurationBuilder configurationBuilder, WebJobsBuilderContext webJobsBuilderContext)
+        {
+            ConfigurationBuilder = configurationBuilder ?? throw new ArgumentNullException(nameof(configurationBuilder));
+            Context = new DefaultFunctionHostBuilderContext(webJobsBuilderContext);
+        }
+
+        public IConfigurationBuilder ConfigurationBuilder { get; private set; }
```
Comment (fabiocav) [inline src/Extensions/DependencyInjection/FunctionsConfigurationBuilder.cs:18]: nit: drop the private setters here

File: src/Extensions/DependencyInjection/FunctionsHostBuilderContext.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+namespace Microsoft.Azure.Functions.Extensions.DependencyInjection
+{
+    /// <summary>
+    /// Context containing the common services and settings for the Functions host.
+    /// </summary>
+    public abstract class FunctionsHostBuilderContext
+    {
+        private readonly WebJobsBuilderContext _webJobsBuilderContext;
+
+        public FunctionsHostBuilderContext(WebJobsBuilderContext webJobsBuilderContext)
```
Comment (fabiocav) [inline src/Extensions/DependencyInjection/FunctionsHostBuilderContext.cs:17]: Is this being renamed from the previous changes?

File: src/Extensions/DependencyInjection/FunctionsStartup.cs [hunk 4]
```diff
@@ -11,21 +10,43 @@ namespace Microsoft.Azure.Functions.Extensions.DependencyInjection
     /// Interface defining a startup configuration action that should be performed
     /// as part of the Functions runtime startup.
     /// </summary>
-    public abstract class FunctionsStartup : IWebJobsStartup
+    public abstract class FunctionsStartup : IWebJobsStartup2, IWebJobsConfigurationStartup
     {
```
Comment (fabiocav) [inline src/Extensions/DependencyInjection/FunctionsStartup.cs:14]: Trying to remember what we decided with the `IWebJobsStartup2`, I thought we had discussed removing it, but it feels like ages ago.

File: src/Extensions/DependencyInjection/FunctionsStartup.cs [hunk 5]
```diff
@@ -11,21 +10,43 @@ namespace Microsoft.Azure.Functions.Extensions.DependencyInjection
...
+
+            ConfigureAppConfiguration(functionsConfigBuilder);
+        }
+
         /// <summary>
         /// Performs the startup configuration action. The runtime will call this
         /// method at the right time during initialization.
         /// </summary>
-        /// <param name="builder">The <see cref="IWebJobsBuilder"/> that can be used to
+        /// <param name="builder">The <see cref="IFunctionsHostBuilder"/> that can be used to
```
Comment (fabiocav) [inline src/Extensions/DependencyInjection/FunctionsStartup.cs:38]: 👍

**Comment by fabiocav** [comment]
A couple of small comments, but we should be able to land this soon! Also, as discussed, the package should be flagged as a preview.

**Comment by luthus** [comment]
Is there an ETA for when a preview package could be available with these changes?

**Comment by brettsam** [comment]
We have to wait for the underlying changes to make it to production or else these won't work. That's likely a couple of weeks (the commits are merged, they'll be going out with the next release train).

**Comment by fabiocav** [comment]
@brettsam we should be able to merge this. Release won't happen until after the core changes are deployed.

**Comment by brettsam** [comment]
True. Will need to update to target 3.0.18 (which is in staging) -- then will submit an update.

**Comment by brettsam** [comment]
Related (but I don't want it to auto-close b/c docs will be done outside this PR): #44

File: src/Extensions/DependencyInjection/FunctionsStartup.cs (hunk 4 above)
Comment (brettsam) [reply src/Extensions/DependencyInjection/FunctionsStartup.cs:14]: It was required -- we needed a new property to hold the Configure() call with a Context.

**Comment by brettsam** [comment]
@fabiocav -- this is ready for final review.

**Comment by hermesramos** [comment]
Slowpokes

File: src/Extensions/DependencyInjection/FunctionsHostBuilderContext.cs (hunk 3 above)
Comment (brettsam) [reply src/Extensions/DependencyInjection/FunctionsHostBuilderContext.cs:17]: Yes it is called `FunctionsHostBuilder` (with an 's') yet this was called `FunctionHostBuilderContext`... so I added the 's' to match it.

**Comment by fabiocav** [comment]
> Slowpokes

🐌

**Comment by jamesharling** [comment]
@brettsam do you have an ETA for when these packages will be published and usable?

**Comment by jamesharling** [comment]
For all subscribers, a preview package has landed in the last few days: https://www.nuget.org/packages/Microsoft.Azure.Functions.Extensions/1.1.0-preview1

Example of the new configuration API in use: https://raw.githubusercontent.com/brettsam/azure-functions-dotnet-extensions/78878f0c724004c5bab129bca5430568f6370a73/src/samples/DependencyInjection/Basic/SampleStartup.cs

**Comment by baharedankoyuncu** [comment]
Looking forward to this, however I found it slightly confusing that the [docs](https://docs.microsoft.com/en-us/azure/azure-functions/functions-dotnet-dependency-injection#customizing-configuration-sources) already mentioned this considering it's a not released yet and requires a preview [package](https://www.nuget.org/packages/Microsoft.Azure.Functions.Extensions/1.1.0-preview1).

**Comment by jamesharling** [comment]
FYI this appears to be formally released: https://www.nuget.org/packages/Microsoft.Azure.Functions.Extensions/1.1.0

**Pull request merged**